src/v/cluster/members_backend.cc

```diff
@@ -569,107 +629,146 @@ void members_backend::calculate_reallocations_batch(
+            // are allocated with new cluster capacity
+            if (meta.update) {
+                if (metadata.get_revision() > meta.update->offset()) {
+                    vlog(
+                      clusterlog.debug,
+                      "skipping reallocating topic {}, its revision {} is "
+                      "greater than node update {}",
+                      tp_ns,
+                      metadata.get_revision(),
+                      meta.update->offset);
+                    continue;
+                }
+            }
 
-            std::erase_if(to_move_from_node, [](const replicas_to_move& v) {
-                return v.left_to_move == 0;
-            });
-
-            std::sort(to_move_from_node.begin(), to_move_from_node.end(), cmp);
-            for (auto& to_move : to_move_from_node) {
-                if (
-                  is_in_replica_set(p.replicas, to_move.id)
-                  && to_move.left_to_move > 0) {
+            for (const auto& p : metadata.get_assignments()) {
+                if (is_in_replica_set(p.replicas, to_move.id)) {
                     partition_reallocation reallocation(
                       model::ntp(tp_ns.ns, tp_ns.tp, p.id), p.replicas.size());
                     reallocation.replicas_to_remove.emplace(to_move.id);
                     auto current_assignment = p;
-                    reassign_replicas(current_assignment, reallocation);
-                    // skip if partition was reassigned to the same node
-                    if (is_reassigned_to_node(reallocation, to_move.id)) {
+                    reassign_replicas(
+                      allocator, current_assignment, reallocation);
+
+                    if (!reallocation.allocation_units.has_value()) {
+                        continue;
+                    }
+                    const auto& new_assignment = reallocation.allocation_units
+                                                   .value()
+                                                   .get_assignments()
+                                                   .begin()
+                                                   ->replicas;
+
+                    auto new_error = error_after_move(
+                      total_replicas,
+                      node_replicas,
+                      allocator,
+                      new_assignment,
+                      p.replicas);
+                    vlog(
+                      clusterlog.trace,
+                      "ntp {} move from {} -> {}, error before {}, error "
+                      "after: {}",
+                      reallocation.ntp,
+                      p.replicas,
+                      new_assignment,
+                      initial_error,
+                      new_error);
+                    /**
+                     * If there is no error improvement, skip this reallocation
+                     */
+                    if (new_error >= initial_error) {
```

Comment:
should we update initial_error here to ensure that every movement improves the error?